src/Runtime/pdf3/internal/generator/browser_session.go

```diff
@@ -0,0 +1,540 @@
+			time.Sleep(testInput.CleanupDelaySeconds * time.Second)
+		}
+
+		// Cleanup browser storage
+		w.cleanupBrowser(req)
+
+		// Retry cleanup if it failed
+		if !req.cleanedUp {
+			log.Printf("[%d, %s] failed to cleanup storage, retrying...\n", w.id, w.currentUrl)
+			for range 3 {
+				w.cleanupBrowser(req)
+				if req.cleanedUp {
+					break
+				}
+			}
+
+			if !req.cleanedUp {
+				log.Fatalf("[%d, %s] failed to cleanup storage, we're in an unsafe state and can't proceed", w.id, w.currentUrl)
+			}
+		}
+		duration := time.Since(start)
+		log.Printf("[%d, %s] cleanup completed in %.2f seconds\n", w.id, w.currentUrl, duration.Seconds())
+	}()
+
+	// Set cookies
+	for _, cookie := range request.Cookies {
+		if req.ctx.Err() != nil {
+			return types.NewPDFError(types.ErrClientDropped, "", req.ctx.Err())
+		}
+
+		sameSite := "Lax"
+		switch cookie.SameSite {
+		case "Strict":
+			sameSite = "Strict"
+		case "None":
+			sameSite = "None"
+		}
+
+		_, err := w.sendCommand("Network.setCookie", map[string]interface{}{
+			"name":     cookie.Name,
+			"value":    cookie.Value,
+			"domain":   cookie.Domain,
+			"path":     "/",
+			"secure":   false,
+			"httpOnly": false,
+			"sameSite": sameSite,
+		})
+		if err != nil {
+			req.tryRespondError(types.NewPDFError(types.ErrSetCookieFail, "", err), w)
+			return nil
+		}
+	}
+
+	// Navigate to URL
+	if req.hasResponded() {
+		return nil
+	}
+	if req.ctx.Err() != nil {
+		req.tryRespondError(types.NewPDFError(types.ErrClientDropped, "", req.ctx.Err()), w)
+		return nil
+	}
+
+	_, err := w.sendCommand("Page.navigate", map[string]interface{}{
+		"url": request.URL,
+	})
+	if err != nil {
+		req.tryRespondError(types.NewPDFError(types.ErrGenerationFail, "", err), w)
+		return nil
+	}
+
+	// Wait for element if specified
+	if request.WaitFor != nil {
+		if req.hasResponded() {
+			return nil
+		}
+		if req.ctx.Err() != nil {
+			req.tryRespondError(types.NewPDFError(types.ErrClientDropped, "", req.ctx.Err()), w)
+			return nil
+		}
+
+		var waitSelector string
+		var timeoutMs int32 = 25000 // default timeout
+
+		if selector, ok := request.WaitFor.AsString(); ok {
+			waitSelector = selector
+		} else if timeout, ok := request.WaitFor.AsTimeout(); ok {
+			time.Sleep(time.Duration(timeout) * time.Millisecond)
+		} else if opts, ok := request.WaitFor.AsOptions(); ok {
+			// Full options with selector, visible, hidden, timeout
+			waitSelector = opts.Selector
+			if opts.Timeout != nil {
+				timeoutMs = *opts.Timeout
+			}
+			// TODO: Implement visible/hidden support if needed
+		}
+
+		// If we have a selector to wait for, use MutationObserver
+		if waitSelector != "" {
+			// Wait for element using MutationObserver via a single Runtime.evaluate with awaitPromise
+			// Match gorod's behavior: if selector is an id ("#id"), use an optimized observer; otherwise use a generic selector observer.
+			var expression string
+			if waitSelector[0] == '#' {
```

Comment:
Technhically, a selector starting with a `#` is not necessarily a pure id. It could e.g. be something like `#id.class` or any other shenanigans afterwards. 

> Technically, the value for an ID attribute may contain any other Unicode character. However, when used in CSS selectors, either from JavaScript using APIs like Document.querySelector or in CSS stylesheets, ID attribute values must be valid [CSS identifiers](https://developer.mozilla.org/en-US/docs/Web/CSS/ident).

A relatively simple pattern could be `^#[\w-]+$`. Although this still excludes ids with escaped characters, and allows ids starting with a number (which is technically not allowed)